lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/HiveAvroTypeBlockHandler.java

```diff
@@ -0,0 +1,442 @@
+import static io.trino.hive.formats.avro.AvroTypeUtils.isSimpleNullableUnion;
+import static io.trino.hive.formats.avro.AvroTypeUtils.throwForCircularReference;
+import static io.trino.hive.formats.avro.AvroTypeUtils.unwrapNullableUnion;
+import static io.trino.hive.formats.avro.BaseAvroTypeBlockHandler.NullBlockBuildingDecoder.INSTANCE;
+import static io.trino.hive.formats.avro.HiveAvroTypeManager.getHiveLogicalVarCharOrCharType;
+import static io.trino.hive.formats.avro.NativeLogicalTypesAvroTypeBlockHandler.getLogicalTypeBuildingFunction;
+import static io.trino.hive.formats.avro.NativeLogicalTypesAvroTypeManager.getAvroLogicalTypeSpiType;
+import static io.trino.hive.formats.avro.NativeLogicalTypesAvroTypeManager.validateLogicalType;
+import static java.time.ZoneOffset.UTC;
+import static java.util.Objects.requireNonNull;
+
+public class HiveAvroTypeBlockHandler
+        extends BaseAvroTypeBlockHandler
+{
+    private final AtomicReference<ZoneId> convertToTimezone = new AtomicReference<>(UTC);
+    private final TimestampType hiveSessionTimestamp;
+
+    public HiveAvroTypeBlockHandler(TimestampType hiveSessionTimestamp)
+    {
+        this.hiveSessionTimestamp = requireNonNull(hiveSessionTimestamp, "hiveSessionTimestamp is null");
+    }
+
+    @Override
+    public void configure(Map<String, byte[]> fileMetadata)
+    {
+        if (fileMetadata.containsKey(AvroHiveConstants.WRITER_TIME_ZONE)) {
+            convertToTimezone.set(ZoneId.of(new String(fileMetadata.get(AvroHiveConstants.WRITER_TIME_ZONE), StandardCharsets.UTF_8)));
+        }
+        else {
+            // legacy path allows this conversion to be skipped with {@link org.apache.hadoop.conf.Configuration} param
+            // currently no way to set that configuration in Trino
+            convertToTimezone.set(TimeZone.getDefault().toZoneId());
+        }
+    }
+
+    @Override
+    public Type typeFor(Schema schema)
+            throws AvroTypeException
+    {
+        throwForCircularReference(schema);
+        if (schema.getType() == Schema.Type.NULL) {
+            // allows of dereference when no base columns from file used
+            // BooleanType chosen rather arbitrarily to be stuffed with null
+            // in response to behavior defined by io.trino.tests.product.hive.TestAvroSchemaStrictness.testInvalidUnionDefaults
+            return BooleanType.BOOLEAN;
+        }
+        else if (schema.getType() == Schema.Type.UNION) {
+            if (isSimpleNullableUnion(schema)) {
+                return typeFor(unwrapNullableUnion(schema));
+            }
+            else {
+                // coerce complex unions to structs
+                return rowTypeForUnion(schema);
+            }
+        }
+        ValidateLogicalTypeResult logicalTypeResult = validateLogicalType(schema);
+        return switch (logicalTypeResult) {
+            case NoLogicalType __ -> super.typeFor(schema);
```

Comment:
I thought about this as an option. The reason I went for this is because there time where the nested types do need to be able to move back up to the parent class. To do with static function we would need to pass in the block handler manually so that it could be invoked, much like I'm doing in the constructor of some of the BlockBuildingDecoders. We can do it that way / we can talk about it more in detail as well. 

We could ever export a visitor pattern as well a see if that makes sense, but I was hoping to use the pattern matching that is now available for my viewing preferences.